File: hello-websocket-go/common/codec.go
package common

import (
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"math"
	"strings"
	"time"
)

// ─── Constants ───────────────────────────────────────────────────────────

const (
	Port            = 9898
	Magic     byte  = 0x48
	Version   byte  = 0x01
	HeaderLen       = 8
	ServerLang      = "GO"
	ClientLang      = "GO"
)

// Message types
const (
	MsgHello            byte = 0x01
	MsgBonjour          byte = 0x02
	MsgEchoRequest      byte = 0x03
	MsgEchoResponse     byte = 0x04
	MsgKissRequest      byte = 0x05
	MsgKissResponse     byte = 0x06
	MsgPing             byte = 0x07
	MsgPong             byte = 0x08
	MsgTimeNotification byte = 0x09
	MsgRandomNumber     byte = 0x0A
	MsgHashResponse     byte = 0x0B
	MsgDisconnect       byte = 0x0C
	MsgError            byte = 0x7F
)

// Error codes
const (
	ErrDecode            int32 = 0x01
	ErrUnknownMsgType   int32 = 0x02
	ErrTruncatedPayload int32 = 0x03
	ErrBadMagic         int32 = 0x04
	ErrBadVersion       int32 = 0x05
	ErrSessionNotFound  int32 = 0x06
	ErrInternal         int32 = 0x07
)

// Intervals
const (
	PingInterval       = 1 * time.Second
	SessionTimeout     = 60 * time.Second
	TimeInterval       = 5 * time.Second
	RandomInterval     = 5 * time.Second
	KissInterval       = 5 * time.Second
)

// ─── Primitive Encoders ──────────────────────────────────────────────────

func WriteU8(buf *[]byte, v uint8) {
	*buf = append(*buf, v)
}

func WriteU16(buf *[]byte, v uint16) {
	var b [2]byte
	binary.BigEndian.PutUint16(b[:], v)
	*buf = append(*buf, b[:]...)
}

func WriteU32(buf *[]byte, v uint32) {
	var b [4]byte
	binary.BigEndian.PutUint32(b[:], v)
	*buf = append(*buf, b[:]...)
}

func WriteI32(buf *[]byte, v int32) {
	WriteU32(buf, uint32(v))
}

func WriteI64(buf *[]byte, v int64) {
	var b [8]byte
	binary.BigEndian.PutUint64(b[:], uint64(v))
	*buf = append(*buf, b[:]...)
}

func WriteString(buf *[]byte, s string) {
	b := []byte(s)
	WriteU32(buf, uint32(len(b)))
	*buf = append(*buf, b...)
}

func WriteKV(buf *[]byte, m map[string]string) {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	WriteU32(buf, uint32(len(keys)))
	for _, k := range keys {
		WriteString(buf, k)
		WriteString(buf, m[k])
	}
}

// ─── Primitive Decoders ──────────────────────────────────────────────────

type Reader struct {
	data []byte
	pos  int
}

func NewReader(data []byte) *Reader {
	return &Reader{data: data}
}

func (r *Reader) ReadU8() (uint8, error) {
	if r.pos+1 > len(r.data) {
		return 0, fmt.Errorf("unexpected end of data at offset %d reading u8", r.pos)
	}
	v := r.data[r.pos]
	r.pos++
	return v, nil
}

func (r *Reader) ReadU16() (uint16, error) {
	if r.pos+2 > len(r.data) {
		return 0, fmt.Errorf("unexpected end of data at offset %d reading u16", r.pos)
	}
	v := binary.BigEndian.Uint16(r.data[r.pos:])
	r.pos += 2
	return v, nil
}

func (r *Reader) ReadU32() (uint32, error) {
	if r.pos+4 > len(r.data) {
		return 0, fmt.Errorf("unexpected end of data at offset %d reading u32", r.pos)
	}
	v := binary.BigEndian.Uint32(r.data[r.pos:])
	r.pos += 4
	return v, nil
}

func (r *Reader) ReadI32() (int32, error) {
	u, err := r.ReadU32()
	return int32(u), err
}

func (r *Reader) ReadI64() (int64, error) {
	if r.pos+8 > len(r.data) {
		return 0, fmt.Errorf("unexpected end of data at offset %d reading i64", r.pos)
	}
	v := int64(binary.BigEndian.Uint64(r.data[r.pos:]))
	r.pos += 8
	return v, nil
}

func (r *Reader) ReadString() (string, error) {
	ln, err := r.ReadU32()
	if err != nil {
		return "", err
	}
	if r.pos+int(ln) > len(r.data) {
		return "", fmt.Errorf("string length %d exceeds remaining data", ln)
	}
	s := string(r.data[r.pos : r.pos+int(ln)])
	r.pos += int(ln)
	return s, nil
}

func (r *Reader) ReadKV() (map[string]string, error) {
	count, err := r.ReadU32()
	if err != nil {
		return nil, err
	}
	m := make(map[string]string, count)
	for i := uint32(0); i < count; i++ {
		k, err := r.ReadString()
		if err != nil {
			return nil, err
		}
		v, err := r.ReadString()
		if err != nil {
			return nil, err
		}
		m[k] = v
	}
	return m, nil
}

// ─── Frame Codec ─────────────────────────────────────────────────────────

func EncodeFrame(msgType byte, payload []byte) []byte {
	buf := make([]byte, HeaderLen+len(payload))
	buf[0] = Magic
	buf[1] = Version
	buf[2] = msgType
	buf[3] = 0x00 // flags
	binary.BigEndian.PutUint32(buf[4:], uint32(len(payload)))
	copy(buf[HeaderLen:], payload)
	return buf
}

func DecodeFrame(data []byte) (msgType byte, payload []byte, err error) {
	if len(data) < HeaderLen {
		return 0, nil, fmt.Errorf("frame too short: %d bytes", len(data))
	}
	if data[0] != Magic {
		return 0, nil, fmt.Errorf("bad magic: 0x%02x", data[0])
	}
	if data[1] != Version {
		return 0, nil, fmt.Errorf("bad version: 0x%02x", data[1])
	}
	msgType = data[2]
	payloadLen := binary.BigEndian.Uint32(data[4:])
	if int(payloadLen) > len(data)-HeaderLen {
		return 0, nil, fmt.Errorf("truncated payload: declared %d, available %d", payloadLen, len(data)-HeaderLen)
	}
	payload = data[HeaderLen : HeaderLen+int(payloadLen)]
	return msgType, payload, nil
}

// ─── Message Types ───────────────────────────────────────────────────────

type Hello struct {
	ClientLanguage string
}

func (m *Hello) Encode() []byte {
	var buf []byte
	WriteString(&buf, m.ClientLanguage)
	return EncodeFrame(MsgHello, buf)
}

func DecodeHello(r *Reader) (*Hello, error) {
	lang, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	return &Hello{ClientLanguage: lang}, nil
}

type Bonjour struct {
	ServerLanguage string
}

func (m *Bonjour) Encode() []byte {
	var buf []byte
	WriteString(&buf, m.ServerLanguage)
	return EncodeFrame(MsgBonjour, buf)
}

func DecodeBonjour(r *Reader) (*Bonjour, error) {
	lang, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	return &Bonjour{ServerLanguage: lang}, nil
}

type EchoRequest struct {
	ID   int64
	Meta string
	Data string
}

func (m *EchoRequest) Encode() []byte {
	var buf []byte
	WriteI64(&buf, m.ID)
	WriteString(&buf, m.Meta)
	WriteString(&buf, m.Data)
	return EncodeFrame(MsgEchoRequest, buf)
}

func DecodeEchoRequest(r *Reader) (*EchoRequest, error) {
	id, err := r.ReadI64()
	if err != nil {
		return nil, err
	}
	meta, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	data, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	return &EchoRequest{ID: id, Meta: meta, Data: data}, nil
}

type EchoResult struct {
	Idx  int64
	Type uint8
	KV   map[string]string
}

type EchoResponse struct {
	Status  int32
	Results []EchoResult
}

func (m *EchoResponse) Encode() []byte {
	var buf []byte
	WriteI32(&buf, m.Status)
	WriteU32(&buf, uint32(len(m.Results)))
	for _, res := range m.Results {
		WriteI64(&buf, res.Idx)
		WriteU8(&buf, res.Type)
		WriteKV(&buf, res.KV)
	}
	return EncodeFrame(MsgEchoResponse, buf)
}

func DecodeEchoResponse(r *Reader) (*EchoResponse, error) {
	status, err := r.ReadI32()
	if err != nil {
		return nil, err
	}
	count, err := r.ReadU32()
	if err != nil {
		return nil, err
	}
	results := make([]EchoResult, 0, count)
	for i := uint32(0); i < count; i++ {
		idx, err := r.ReadI64()
		if err != nil {
			return nil, err
		}
		typ, err := r.ReadU8()
		if err != nil {
			return nil, err
		}
		kv, err := r.ReadKV()
		if err != nil {
			return nil, err
		}
		results = append(results, EchoResult{Idx: idx, Type: typ, KV: kv})
	}
	return &EchoResponse{Status: status, Results: results}, nil
}

type KissRequest struct {
	OSName         string
	OSVersion      string
	OSRelease      string
	OSArchitecture string
}

func (m *KissRequest) Encode() []byte {
	var buf []byte
	WriteString(&buf, m.OSName)
	WriteString(&buf, m.OSVersion)
	WriteString(&buf, m.OSRelease)
	WriteString(&buf, m.OSArchitecture)
	return EncodeFrame(MsgKissRequest, buf)
}

func DecodeKissRequest(r *Reader) (*KissRequest, error) {
	name, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	ver, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	rel, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	arch, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	return &KissRequest{OSName: name, OSVersion: ver, OSRelease: rel, OSArchitecture: arch}, nil
}

type KissResponse struct {
	Language string
	Encoding string
	TimeZone string
}

func (m *KissResponse) Encode() []byte {
	var buf []byte
	WriteString(&buf, m.Language)
	WriteString(&buf, m.Encoding)
	WriteString(&buf, m.TimeZone)
	return EncodeFrame(MsgKissResponse, buf)
}

func DecodeKissResponse(r *Reader) (*KissResponse, error) {
	lang, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	enc, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	tz, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	return &KissResponse{Language: lang, Encoding: enc, TimeZone: tz}, nil
}

type Ping struct {
	TimestampMs int64
}

func (m *Ping) Encode() []byte {
	var buf []byte
	WriteI64(&buf, m.TimestampMs)
	return EncodeFrame(MsgPing, buf)
}

func DecodePing(r *Reader) (*Ping, error) {
	ts, err := r.ReadI64()
	if err != nil {
		return nil, err
	}
	return &Ping{TimestampMs: ts}, nil
}

type Pong struct {
	TimestampMs int64
}

func (m *Pong) Encode() []byte {
	var buf []byte
	WriteI64(&buf, m.TimestampMs)
	return EncodeFrame(MsgPong, buf)
}

func DecodePong(r *Reader) (*Pong, error) {
	ts, err := r.ReadI64()
	if err != nil {
		return nil, err
	}
	return &Pong{TimestampMs: ts}, nil
}

type TimeNotification struct {
	TimestampMs int64
	ISO8601     string
}

func (m *TimeNotification) Encode() []byte {
	var buf []byte
	WriteI64(&buf, m.TimestampMs)
	WriteString(&buf, m.ISO8601)
	return EncodeFrame(MsgTimeNotification, buf)
}

func DecodeTimeNotification(r *Reader) (*TimeNotification, error) {
	ts, err := r.ReadI64()
	if err != nil {
		return nil, err
	}
	iso, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	return &TimeNotification{TimestampMs: ts, ISO8601: iso}, nil
}

type RandomNumber struct {
	ID     int64
	Number int64
}

func (m *RandomNumber) Encode() []byte {
	var buf []byte
	WriteI64(&buf, m.ID)
	WriteI64(&buf, m.Number)
	return EncodeFrame(MsgRandomNumber, buf)
}

func DecodeRandomNumber(r *Reader) (*RandomNumber, error) {
	id, err := r.ReadI64()
	if err != nil {
		return nil, err
	}
	num, err := r.ReadI64()
	if err != nil {
		return nil, err
	}
	return &RandomNumber{ID: id, Number: num}, nil
}

type HashResponse struct {
	ID      int64
	HashHex string
}

func (m *HashResponse) Encode() []byte {
	var buf []byte
	WriteI64(&buf, m.ID)
	WriteString(&buf, m.HashHex)
	return EncodeFrame(MsgHashResponse, buf)
}

func DecodeHashResponse(r *Reader) (*HashResponse, error) {
	id, err := r.ReadI64()
	if err != nil {
		return nil, err
	}
	hash, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	return &HashResponse{ID: id, HashHex: hash}, nil
}

type Disconnect struct {
	Reason string
}

func (m *Disconnect) Encode() []byte {
	var buf []byte
	WriteString(&buf, m.Reason)
	return EncodeFrame(MsgDisconnect, buf)
}

func DecodeDisconnect(r *Reader) (*Disconnect, error) {
	reason, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	return &Disconnect{Reason: reason}, nil
}

type ErrorMsg struct {
	Code    int32
	Message string
}

func (m *ErrorMsg) Encode() []byte {
	var buf []byte
	WriteI32(&buf, m.Code)
	WriteString(&buf, m.Message)
	return EncodeFrame(MsgError, buf)
}

func DecodeErrorMsg(r *Reader) (*ErrorMsg, error) {
	code, err := r.ReadI32()
	if err != nil {
		return nil, err
	}
	msg, err := r.ReadString()
	if err != nil {
		return nil, err
	}
	return &ErrorMsg{Code: code, Message: msg}, nil
}

// ─── Message Dispatch ────────────────────────────────────────────────────

type Message struct {
	Type     byte
	Hello    *Hello
	Bonjour  *Bonjour
	EchoReq  *EchoRequest
	EchoResp *EchoResponse
	KissReq  *KissRequest
	KissResp *KissResponse
	Ping     *Ping
	Pong     *Pong
	TimeNotif *TimeNotification
	Random   *RandomNumber
	Hash     *HashResponse
	Disconnect *Disconnect
	Error    *ErrorMsg
}

func DecodeMessage(data []byte) (*Message, error) {
	msgType, payload, err := DecodeFrame(data)
	if err != nil {
		return nil, err
	}
	r := NewReader(payload)
	msg := &Message{Type: msgType}
	switch msgType {
	case MsgHello:
		msg.Hello, err = DecodeHello(r)
	case MsgBonjour:
		msg.Bonjour, err = DecodeBonjour(r)
	case MsgEchoRequest:
		msg.EchoReq, err = DecodeEchoRequest(r)
	case MsgEchoResponse:
		msg.EchoResp, err = DecodeEchoResponse(r)
	case MsgKissRequest:
		msg.KissReq, err = DecodeKissRequest(r)
	case MsgKissResponse:
		msg.KissResp, err = DecodeKissResponse(r)
	case MsgPing:
		msg.Ping, err = DecodePing(r)
	case MsgPong:
		msg.Pong, err = DecodePong(r)
	case MsgTimeNotification:
		msg.TimeNotif, err = DecodeTimeNotification(r)
	case MsgRandomNumber:
		msg.Random, err = DecodeRandomNumber(r)
	case MsgHashResponse:
		msg.Hash, err = DecodeHashResponse(r)
	case MsgDisconnect:
		msg.Disconnect, err = DecodeDisconnect(r)
	case MsgError:
		msg.Error, err = DecodeErrorMsg(r)
	default:
		return nil, fmt.Errorf("unknown message type: 0x%02x", msgType)
	}
	if err != nil {
		return nil, err
	}
	return msg, nil
}

// ─── Utility ─────────────────────────────────────────────────────────────

func NowMs() int64 {
	return time.Now().UnixMilli()
}

func NowISO() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05Z")
}

func HashNumber(num int64) string {
	h := sha256.Sum256([]byte(fmt.Sprintf("%d", num)))
	return fmt.Sprintf("%x", h[:5])
}

func Log(name, msg string) {
	ts := time.Now().Format("2006-01-02 15:04:05")
	fmt.Printf("[%s] [INFO] [%s] %s\n", ts, name, msg)
}

func Logf(name, format string, args ...interface{}) {
	Log(name, fmt.Sprintf(format, args...))
}

// FloatToBits and back (unused but available for future extensions)
func Float64ToBits(f float64) uint64 {
	return math.Float64bits(f)
}

func BitsToFloat64(b uint64) float64 {
	return math.Float64frombits(b)
}

// SafeI64ToString converts an int64 to string without using strconv (for minimal deps)
func I64ToString(v int64) string {
	return fmt.Sprintf("%d", v)
}

// JoinStrings joins strings with a separator
func JoinStrings(parts []string, sep string) string {
	return strings.Join(parts, sep)
}

File: hello-websocket-nodejs/common/codec.js
'use strict';

const crypto = require('crypto');
const os = require('os');
const { TextDecoder } = require('util');

// Strict UTF-8 decoder: throws on malformed byte sequences instead of
// silently substituting U+FFFD (the default Buffer#toString('utf-8')
// behavior), so ReadString enforces PROTOCOL.md §3's "valid UTF-8" rule.
const UTF8_DECODER = new TextDecoder('utf-8', { fatal: true });

// ─── Constants ───────────────────────────────────────────────────────────

const PORT = 9898;
const MAGIC = 0x48;
const VERSION = 0x01;
const HEADER_LEN = 8;
const SERVER_LANG = 'NODEJS';
const CLIENT_LANG = 'NODEJS';

const MSG_HELLO = 0x01;
const MSG_BONJOUR = 0x02;
const MSG_ECHO_REQUEST = 0x03;
const MSG_ECHO_RESPONSE = 0x04;
const MSG_KISS_REQUEST = 0x05;
const MSG_KISS_RESPONSE = 0x06;
const MSG_PING = 0x07;
const MSG_PONG = 0x08;
const MSG_TIME_NOTIFICATION = 0x09;
const MSG_RANDOM_NUMBER = 0x0A;
const MSG_HASH_RESPONSE = 0x0B;
const MSG_DISCONNECT = 0x0C;
const MSG_ERROR = 0x7F;

const ERR_DECODE = 0x01;
const ERR_UNKNOWN_MSG_TYPE = 0x02;
const ERR_TRUNCATED_PAYLOAD = 0x03;
const ERR_BAD_MAGIC = 0x04;
const ERR_BAD_VERSION = 0x05;
const ERR_SESSION_NOT_FOUND = 0x06;
const ERR_INTERNAL = 0x07;

const PING_INTERVAL = 1000;
const SESSION_TIMEOUT = 60000;
const TIME_INTERVAL = 5000;
const RANDOM_INTERVAL = 5000;
const KISS_INTERVAL = 5000;

// ─── ByteWriter ─────────────────────────────────────────────────────────

class ByteWriter {
    constructor() { this.buf = []; }

    writeU8(v) { this.buf.push(v & 0xFF); }
    writeU16(v) { this.buf.push((v >> 8) & 0xFF, v & 0xFF); }
    writeU32(v) { this.buf.push((v >>> 24) & 0xFF, (v >> 16) & 0xFF, (v >> 8) & 0xFF, v & 0xFF); }
    writeI32(v) { this.writeU32(v >>> 0); }
    writeI64(v) {
        const n = BigInt.asUintN(64, BigInt(v));
        this.writeU32(Number((n >> 32n) & 0xFFFFFFFFn));
        this.writeU32(Number(n & 0xFFFFFFFFn));
    }
    writeString(s) {
        const b = Buffer.from(s, 'utf-8');
        this.writeU32(b.length);
        for (const byte of b) this.buf.push(byte);
    }
    writeKV(m) {
        const keys = Object.keys(m);
        this.writeU32(keys.length);
        for (const k of keys) { this.writeString(k); this.writeString(m[k]); }
    }
    data() { return Buffer.from(this.buf); }
}

// ─── ByteReader ─────────────────────────────────────────────────────────

class ByteReader {
    constructor(data) { this.data = data; this.pos = 0; }

    remaining() { return this.data.length - this.pos; }
    ensure(n, field) {
        if (!Number.isSafeInteger(n) || n < 0 || n > this.remaining()) {
            throw new Error(`${field} requires ${n} bytes, only ${this.remaining()} remain`);
        }
    }
    readU8() { this.ensure(1, 'u8'); return this.data[this.pos++]; }
    readU16() { this.ensure(2, 'u16'); const v = (this.data[this.pos] << 8) | this.data[this.pos + 1]; this.pos += 2; return v; }
    readU32() {
        this.ensure(4, 'u32');
        const v = (this.data[this.pos] * 0x1000000) + (this.data[this.pos + 1] << 16) +
                  (this.data[this.pos + 2] << 8) + this.data[this.pos + 3];
        this.pos += 4;
        return v;
    }
    readI32() { return this.readU32() | 0; }
    readI64() {
        const hi = BigInt(this.readU32());
        const lo = BigInt(this.readU32());
        return BigInt.asIntN(64, (hi << 32n) | lo);
    }
    readString() {
        const ln = this.readU32();
        this.ensure(ln, 'string');
        const bytes = this.data.subarray(this.pos, this.pos + ln);
        // PROTOCOL.md §3 requires string payloads to be valid UTF-8 bytes.
        // Buffer#toString('utf-8') silently replaces invalid sequences with
        // U+FFFD instead of rejecting them; use a fatal TextDecoder so
        // malformed input is rejected like the Go/Rust/Python/C++ decoders.
        let s;
        try {
            s = UTF8_DECODER.decode(bytes);
        } catch {
            throw new Error('string payload is not valid UTF-8');
        }
        this.pos += ln;
        return s;
    }
    readKV() {
        const count = this.readU32();
        if (count > Math.floor(this.remaining() / 8)) throw new Error(`kv count ${count} exceeds remaining payload`);
        const m = {};
        for (let i = 0; i < count; i++) {
            const k = this.readString();
            const v = this.readString();
            m[k] = v;
        }
        return m;
    }
}

// ─── Frame Codec ─────────────────────────────────────────────────────────

function encodeFrame(msgType, payload) {
    const header = Buffer.alloc(HEADER_LEN);
    header[0] = MAGIC;
    header[1] = VERSION;
    header[2] = msgType;
    header[3] = 0x00;
    header.writeUInt32BE(payload.length, 4);
    return Buffer.concat([header, payload]);
}

function decodeFrame(data) {
    if (data.length < HEADER_LEN) throw new Error(`frame too short: ${data.length}`);
    if (data[0] !== MAGIC) throw new Error(`bad magic: 0x${data[0].toString(16)}`);
    if (data[1] !== VERSION) throw new Error(`bad version: 0x${data[1].toString(16)}`);
    const msgType = data[2];
    const payloadLen = data.readUInt32BE(4);
    if (payloadLen !== data.length - HEADER_LEN) {
        throw new Error(`payload length mismatch: declared ${payloadLen}, available ${data.length - HEADER_LEN}`);
    }
    return { msgType, payload: data.slice(HEADER_LEN, HEADER_LEN + payloadLen) };
}

// ─── Message Encoders ────────────────────────────────────────────────────

function encodeHello(clientLanguage) {
    const w = new ByteWriter();
    w.writeString(clientLanguage);
    return encodeFrame(MSG_HELLO, w.data());
}

function encodeBonjour(serverLanguage) {
    const w = new ByteWriter();
    w.writeString(serverLanguage);
    return encodeFrame(MSG_BONJOUR, w.data());
}

function encodeEchoRequest(id, meta, data) {
    const w = new ByteWriter();
    w.writeI64(id);
    w.writeString(meta);
    w.writeString(data);
    return encodeFrame(MSG_ECHO_REQUEST, w.data());
}

function encodeEchoResponse(status, results) {
    const w = new ByteWriter();
    w.writeI32(status);
    w.writeU32(results.length);
    for (const r of results) {
        w.writeI64(r.idx);
        w.writeU8(r.type);
        w.writeKV(r.kv);
    }
    return encodeFrame(MSG_ECHO_RESPONSE, w.data());
}

function encodeKissRequest(osName, osVersion, osRelease, osArch) {
    const w = new ByteWriter();
    w.writeString(osName); w.writeString(osVersion);
    w.writeString(osRelease); w.writeString(osArch);
    return encodeFrame(MSG_KISS_REQUEST, w.data());
}

function encodeKissResponse(language, encoding, timeZone) {
    const w = new ByteWriter();
    w.writeString(language); w.writeString(encoding); w.writeString(timeZone);
    return encodeFrame(MSG_KISS_RESPONSE, w.data());
}

function encodePing(ts) {
    const w = new ByteWriter();
    w.writeI64(ts);
    return encodeFrame(MSG_PING, w.data());
}

function encodePong(ts) {
    const w = new ByteWriter();
    w.writeI64(ts);
    return encodeFrame(MSG_PONG, w.data());
}

function encodeTimeNotification(ts, iso) {
    const w = new ByteWriter();
    w.writeI64(ts);
    w.writeString(iso);
    return encodeFrame(MSG_TIME_NOTIFICATION, w.data());
}

function encodeRandomNumber(id, number) {
    const w = new ByteWriter();
    w.writeI64(id);
    w.writeI64(number);
    return encodeFrame(MSG_RANDOM_NUMBER, w.data());
}

function encodeHashResponse(id, hashHex) {
    const w = new ByteWriter();
    w.writeI64(id);
    w.writeString(hashHex);
    return encodeFrame(MSG_HASH_RESPONSE, w.data());
}

function encodeDisconnect(reason) {
    const w = new ByteWriter();
    w.writeString(reason);
    return encodeFrame(MSG_DISCONNECT, w.data());
}

function encodeError(code, message) {
    const w = new ByteWriter();
    w.writeI32(code);
    w.writeString(message);
    return encodeFrame(MSG_ERROR, w.data());
}

// ─── Message Decoder ────────────────────────────────────────────────────

function decodeMessage(data) {
    const { msgType, payload } = decodeFrame(data);
    const r = new ByteReader(payload);
    const msg = { type: msgType };

    switch (msgType) {
        case MSG_HELLO:
            msg.hello = { clientLanguage: r.readString() };
            break;
        case MSG_BONJOUR:
            msg.bonjour = { serverLanguage: r.readString() };
            break;
        case MSG_ECHO_REQUEST:
            msg.echoReq = { id: r.readI64(), meta: r.readString(), data: r.readString() };
            break;
        case MSG_ECHO_RESPONSE:
            msg.echoResp = { status: r.readI32(), results: [] };
            const count = r.readU32();
            if (count > Math.floor(r.remaining() / 13)) throw new Error(`result count ${count} exceeds remaining payload`);
            for (let i = 0; i < count; i++) {
                msg.echoResp.results.push({ idx: r.readI64(), type: r.readU8(), kv: r.readKV() });
            }
            break;
        case MSG_KISS_REQUEST:
            msg.kissReq = { osName: r.readString(), osVersion: r.readString(), osRelease: r.readString(), osArchitecture: r.readString() };
            break;
        case MSG_KISS_RESPONSE:
            msg.kissResp = { language: r.readString(), encoding: r.readString(), timeZone: r.readString() };
            break;
        case MSG_PING:
            msg.ping = { timestampMs: r.readI64() };
            break;
        case MSG_PONG:
            msg.pong = { timestampMs: r.readI64() };
            break;
        case MSG_TIME_NOTIFICATION:
            msg.timeNotif = { timestampMs: r.readI64(), iso8601: r.readString() };
            break;
        case MSG_RANDOM_NUMBER:
            msg.random = { id: r.readI64(), number: r.readI64() };
            break;
        case MSG_HASH_RESPONSE:
            msg.hash = { id: r.readI64(), hashHex: r.readString() };
            break;
        case MSG_DISCONNECT:
            msg.disconnect = { reason: r.readString() };
            break;
        case MSG_ERROR:
            msg.error = { code: r.readI32(), message: r.readString() };
            break;
        default:
            throw new Error(`unknown message type: 0x${msgType.toString(16)}`);
    }
    return msg;
}

// ─── Utility ─────────────────────────────────────────────────────────────

function nowMs() { return Date.now(); }

function nowISO() { return new Date().toISOString().replace(/\.\d+Z$/, 'Z'); }

function hashNumber(num) {
    const h = crypto.createHash('sha256').update(String(num)).digest('hex');
    return h.substring(0, 10);
}

function log(name, msg) {
    const ts = new Date().toISOString().replace('T', ' ').substring(0, 19);
    console.log(`[${ts}] [INFO] [${name}] ${msg}`);
}

module.exports = {
    PORT, MAGIC, VERSION, HEADER_LEN, SERVER_LANG, CLIENT_LANG,
    MSG_HELLO, MSG_BONJOUR, MSG_ECHO_REQUEST, MSG_ECHO_RESPONSE,
    MSG_KISS_REQUEST, MSG_KISS_RESPONSE, MSG_PING, MSG_PONG,
    MSG_TIME_NOTIFICATION, MSG_RANDOM_NUMBER, MSG_HASH_RESPONSE,
    MSG_DISCONNECT, MSG_ERROR,
    ERR_DECODE, ERR_UNKNOWN_MSG_TYPE, ERR_TRUNCATED_PAYLOAD,
    ERR_BAD_MAGIC, ERR_BAD_VERSION, ERR_SESSION_NOT_FOUND, ERR_INTERNAL,
    PING_INTERVAL, SESSION_TIMEOUT, TIME_INTERVAL, RANDOM_INTERVAL, KISS_INTERVAL,
    ByteWriter, ByteReader, encodeFrame, decodeFrame,
    encodeHello, encodeBonjour, encodeEchoRequest, encodeEchoResponse,
    encodeKissRequest, encodeKissResponse, encodePing, encodePong,
    encodeTimeNotification, encodeRandomNumber, encodeHashResponse,
    encodeDisconnect, encodeError, decodeMessage,
    nowMs, nowISO, hashNumber, log
};
